clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -120,12 +116,18 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
         changeSubscription(topics);
     }
 
-    public void changeSubscription(Collection<String> topicsToSubscribe) {
-        if (!this.subscription.equals(new HashSet<>(topicsToSubscribe))) {
-            this.subscription.clear();
-            this.subscription.addAll(topicsToSubscribe);
+    public void subscribeFromPattern(Set<String> topics) {
+        if (subscriptionType != SubscriptionType.AUTO_PATTERN)
+            throw new IllegalArgumentException("Attempt to subscribe from pattern while subscription type set to " +
+                    subscriptionType);
```

Comment:
@hachikuji Does this error indicate a different issue that we would normally report using the defined `SUBSCRIPTION_EXCEPTION_MESSAGE` in other methods? In `subscribe(pattern, listener)` method we already make sure that `subscriptionType` is properly set, so I'm guessing what you added above helps with the race condition fix. But it would be great if you can confirm. Thanks.